### ISSUE TITLE
[HOLD for payment 2023-01-23] [$4000] mWeb - Copy to email/URL is not working correctly - reported by @mateusbra

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#8311]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___
**Issue was found when executing #8195**
## Action Performed:
1. Log in on New Expensify.
2. Open a report.
3. Type an URL and an e-mail on chat.

## Expected Result:
The message "Copy e-mail to clipboard" displayed when you try to copy an email and "copy URL to clipboard" tapping URL

## Actual Result:
The message "Copy to clipboard" displayed when you try to copy an email. Sometimes it shows as regular context menu

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Mobile Web

**Version Number:** 1.1.46-0

**Reproducible in staging?:**  y

**Reproducible in production?:** y (New feature)

**Email or phone of affected tester (no customers):** 

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/160028142-3fe96a20-6e66-4605-ae2a-618bc322f957.mp4

https://user-images.githubusercontent.com/43996225/160028204-1fe2a591-3466-430e-9802-147d0baa43ee.mp4

**Expensify/Expensify Issue URL:** 

**Issue reported by:** Reported by @mateusbra https://expensify.slack.com/archives/C01GTK53T8Q/p1647474115298459

**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)
**Issue was found when executing #8195**
## Upwork Automation - Do Not Edit
- Upwork Job URL: https://www.upwork.com/jobs/~01a8294a9c054ef83d
- Upwork Job ID: 1595258571112300544

**Comment by mateusbra** [comment]
@kavimuru I found this issue while working on https://github.com/Expensify/App/pull/8195 and reported on slack here: https://app.slack.com/client/T03SC9DTT/C01GTK53T8Q does that count for reporting bounty?

**Comment by kavimuru** [comment]
@mateusbra  I don't have access to the slack. But as per this comment, https://github.com/Expensify/App/pull/8195#issuecomment-1070240184 yes. Also tester can repro this bug too https://github.com/Expensify/App/issues/8006#issuecomment-1069334038

**Comment by johnmlee101** [comment]
Opening this up!

**Comment by johnmlee101** [comment]
This should stay daily priority

**Comment by michaelhaxhiu** [comment]
@mateusbra do you have interest in working on this one sice you discovered it?

@kavimuru can you please make sure we at the "@ reported by" to a GH title if they are due the bug reporting bounty? Seems applicable in this case for @mateusbra.

**Comment by Christinadobrzyn** [comment]
Posted job to Upwork;

Internal job - https://www.upwork.com/ab/applicants/1508625903649284096/job-details
External job - https://www.upwork.com/jobs/~01bd32eda238a19082

**Comment by ZamarLancer** [comment]
Hi,
This is Zamar from Upwork.com
I will fix the issue by implementing URL parsing feature in ContextMenuAction.
Now there is no checking parts for url/email when the context menu is shown.
So there will be a validation for the clipboard text, which will check if it is an email address or an url; and switch the caption of the menu item according to the result.
Thank you.

**Comment by parasharrajat** [comment]
I don't see how this is different from https://github.com/Expensify/App/issues/8006. This should have been fixed on https://github.com/Expensify/App/issues/8311. cc: @Santhosh-Sellavel. 

why do we need a different issue?

**Comment by Santhosh-Sellavel** [comment]
@parasharrajat 
#8195  - addresses the issue where clicking on an email in the message showed the `copy URL` option instead of `copy email`.
https://github.com/Expensify/App/issues/8311  - But the existing copy URL itself not working on mWeb at all.

**Comment by marcochavezf** [comment]
Not overdue, waiting for proposals.

**Comment by Christinadobrzyn** [comment]
I'm going to be ooo until April 26th so I'm going to assign a new CM to this issue. Thank you!

**Comment by Christinadobrzyn** [comment]
Price increase to $500 

waiting on prosposals...

**Comment by eVoloshchak** [comment]
**Root cause:**
When we long press on an email/link,[ BaseAnchorForCommentsOnly's](https://github.com/Expensify/App/blob/main/src/components/AnchorForCommentsOnly/BaseAnchorForCommentsOnly/index.js) [onSecondaryInteraction](https://github.com/Expensify/App/blob/6c53be1ea7a022e8a5e788cc7e7efd62c4573474/src/components/AnchorForCommentsOnly/BaseAnchorForCommentsOnly/index.js#L62) is fired, calling `ReportActionContextMenu.showContextMenu` with a corresponding type (`CONTEXT_MENU_TYPES.EMAIL` or `CONTEXT_MENU_TYPES.LINK`).
Howewer, since `BaseAnchorForCommentsOnly` is wrapped inside a second `PressableWithSecondaryInteraction` on [ReportActionItem](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionItem.js) page, it's [showPopover](https://github.com/Expensify/App/blob/6c53be1ea7a022e8a5e788cc7e7efd62c4573474/src/pages/home/report/ReportActionItem.js#L99) method is also fired, calling `ReportActionContextMenu.showContextMenu` with a `CONTEXT_MENU_TYPES.REPORT_ACTION` type.

So, in short, when we long press, `ReportActionContextMenu.showContextMenu` is called with the right type and then almost instantly it gets called with the wrong type, which results in the wrong context menu being shown. 

**Proposal:**
Throttle [showContextMenu](https://github.com/Expensify/App/blob/6c53be1ea7a022e8a5e788cc7e7efd62c4573474/src/pages/home/report/ContextMenu/PopoverReportActionContextMenu.js#L114) method, so the second call get's omitted. 
```javascript
  showContextMenu = _.throttle((
      type,
      event,
      selection,
      contextMenuAnchor,
      reportID,
      reportAction,
      draftMessage,
      onShow = () => {},
      onHide = () => {},
  ) => {
      const nativeEvent = event.nativeEvent || {};
      this.contextMenuAnchor = contextMenuAnchor;

      // Singleton behaviour of ContextMenu creates race conditions when user requests multiple contextMenus.
      // But it is possible that every new request registers new callbacks thus instanceID is used to corelate those callbacks
      this.instanceID = Math.random().toString(36).substr(2, 5);

      // Register the onHide callback only when Popover is shown to remove the race conditions when there are mutltiple popover open requests
      this.onPopoverShow = () => {
          onShow();
          this.onPopoverHide = onHide;
      };
      this.getContextMenuMeasuredLocation().then(({x, y}) => {
          this.setState({
              cursorRelativePosition: {
                  horizontal: nativeEvent.pageX - x,
                  vertical: nativeEvent.pageY - y,
              },
              popoverAnchorPosition: {
                  horizontal: nativeEvent.pageX,
                  vertical: nativeEvent.pageY,
              },
              type,
              reportID,
              reportAction,
              selection,
              isPopoverVisible: true,
              reportActionDraftMessage: draftMessage,
          });
      });
  }, 100, {trailing: false})
```

**Result:**

Chrome on Android


https://user-images.githubusercontent.com/9059945/163683498-971dc82d-ae30-49a4-93df-ae03d8673527.mp4



Safari on iOS


https://user-images.githubusercontent.com/9059945/163683471-88facc87-5664-4af5-b166-934eddc856e3.mp4

**Comment by parasharrajat** [comment]
Root cause seems correct but I don't like the proposal.

**Comment by eVoloshchak** [comment]
> Root cause seems correct but I don't like the proposal.

I've found a simpler fix:
The issue arises because in [PressableWithSecondaryInteraction/index.js](https://github.com/Expensify/App/tree/main/src/components/PressableWithSecondaryInteraction/index.js), we have a LongPressGestureHandler, which wraps Pressable, so essentially links are wrapped like this:
```javascript
<LongPressGestureHandler>
  <Pressable>
    <LongPressGestureHandler>
      <Pressable>
        <TextOfTheLink>
      </Pressable>
    </LongPressGestureHandler>
  </Pressable>
</LongPressGestureHandler>
```
To fix this, we need to:

1. Add `longPressGestureHandlerEnabled={false}` to [this](https://github.com/Expensify/App/blob/6c53be1ea7a022e8a5e788cc7e7efd62c4573474/src/pages/home/report/ReportActionItem.js#L164) line on `ReportActionItem` (prop probably needs a better name)
2. Change [this](https://github.com/Expensify/App/blob/6c53be1ea7a022e8a5e788cc7e7efd62c4573474/src/components/PressableWithSecondaryInteraction/index.js#L34) line to `if (e.nativeEvent.state !== State.ACTIVE || !this.props.longPressGestureHandlerEnabled) {`

To be honest, I'm not entirely sure **why** it works, but it does

https://user-images.githubusercontent.com/9059945/163777528-d0ace458-b6e2-40f8-91c7-534dce3f4870.mp4

**Comment by parasharrajat** [comment]
Ok. What is the base of this solution? Why do you think we should/have to do this? 

The proposal is still lacking the proper explanation of Why the solution?

**Comment by eVoloshchak** [comment]
> Ok. What is the base of this solution? Why do you think we should/have to do this?
> 
> The proposal is still lacking the proper explanation of Why the solution?

Case 1: user long presses on a comment: [LongPressGestureHandler's](https://github.com/Expensify/App/blob/6c53be1ea7a022e8a5e788cc7e7efd62c4573474/src/components/PressableWithSecondaryInteraction/index.js#L69) `onHandlerStateChange` fires, which in turn calls [onSecondaryInteraction={this.showPopover}](https://github.com/Expensify/App/blob/6c53be1ea7a022e8a5e788cc7e7efd62c4573474/src/pages/home/report/ReportActionItem.js#L158)

Case 2: user long presses on a link inside the comment: [LongPressGestureHandler's](https://github.com/Expensify/App/blob/6c53be1ea7a022e8a5e788cc7e7efd62c4573474/src/components/PressableWithSecondaryInteraction/index.js#L69) `onHandlerStateChange` fires, which in turn calls [onSecondaryInteraction={(e) => ReportActionContextMenu.showContextMenu()}](https://github.com/Expensify/App/blob/6c53be1ea7a022e8a5e788cc7e7efd62c4573474/src/components/AnchorForCommentsOnly/BaseAnchorForCommentsOnly/index.js#L64). 
But, since in this case we have one `LongPressGestureHandler's` nested inside another one, both of their onHandlerStateChange events are fired, one on `BaseAnchorForCommentsOnly/index.js ` (the right one) and one from `ReportActionItem.js` (the wrong one).

Adding `longPressGestureHandlerEnabled={false}` will essentially disable parent's `onHandlerStateChange` in case it's child has already fired `onHandlerStateChange`.

**Comment by parasharrajat** [comment]
I understood the root cause. I got your solution and how this works. But what I am asking is why your solution(Workaround)?

Why do you think we should do either of your changes? Is there any systematic way of preventing event bubbling?  

--- 

I think there is a better solution. I will wait to see that proposal before making the choice.

**Comment by eVoloshchak** [comment]
> I understood the root cause. I got your solution and how this works. But what I am asking is why your solution(Workaround)?
> 
> Why do you think we should do either of your changes? Is there any systematic way of preventing event bubbling?

Got it, sorry for the misunderstanding.
I'll try to find a more universal approach a bit later

**Comment by marcochavezf** [comment]
Not overdue, waiting for proposals.

**Comment by arielgreen** [comment]
[Price increased.](https://expensify.slack.com/archives/C01GTK53T8Q/p1651011373563439)

**Comment by b1tjoy** [comment]
### Proposal
### Reason
When we tap the email address, the inner PressableWithSecondaryInteraction component fire an event, in the event handler we call ReportActionContextMenu.showContextMenu() with type 'EMAIL':
https://github.com/Expensify/App/blob/4c64bec3f38af925a01911155eab54feafc96cda/src/components/AnchorForCommentsOnly/BaseAnchorForCommentsOnly/index.js#L64-L76
And the outer PressableWithSecondaryInteraction component also fire an event after that, in the event handler we call ReportActionContextMenu.showContextMenu() with type 'REPORT_ACTION':
https://github.com/Expensify/App/blob/4c64bec3f38af925a01911155eab54feafc96cda/src/pages/home/report/ReportActionItem.js#L99-L115
The outer component toggle the modal popover with regular context menu 'Copy to clipboard', so we can't see the context menu 'Copy e-mail to clipboard'.
### Solution 1
Since the inner PressableWithSecondaryInteraction already toggle the modal popover, and isPopoverVisible property stored in PopoverReportActionContextMenu's state, we can check isPopoverVisible is true in showContextMenu function, if it's true, do nothing and return.
https://github.com/Expensify/App/blob/68f0bbf9c7fe8a95c7d65b0a9153229375863d1e/src/pages/home/report/ContextMenu/PopoverReportActionContextMenu.js#L114-L124
`+       if (this.state.isPopoverVisible) return;`
https://github.com/Expensify/App/blob/68f0bbf9c7fe8a95c7d65b0a9153229375863d1e/src/pages/home/report/ContextMenu/PopoverReportActionContextMenu.js#L125-L126

**Comment by parasharrajat** [comment]
Hmm! I strongly recommend looking back at the previous discussion. I asked

> Why do you think we should do either of your changes? Is there any systematic way of preventing event bubbling?

@b1tjoy your proposal looks hacky to me.

**Comment by arielgreen** [comment]
[Price increased.](https://expensify.slack.com/archives/C01GTK53T8Q/p1651621050549769)

**Comment by eVoloshchak** [comment]
**Proposal for mWeb:**

1. This is not a problem with the Expensify App, it's an inconsistency of `react-native-gesture-handler`. [This](https://snack.expo.dev/rLAPULxbW) is a minimal reproduction snack, running it on web you can see that when longpressing on text, two console statements appear, while there's should be only one ('pressed 2')
2. Since issue is occuring only on `mWeb`, we're interested in  [PressableWithSecondaryInteraction/index.js](https://github.com/Expensify/App/blob/main/src/components/PressableWithSecondaryInteraction/index.js) file. 
LongPressGestureHandler was added to this file in [this PR](https://github.com/Expensify/App/pull/8018), which resolved [#7462](https://github.com/Expensify/App/issues/7462). However, [#7462](https://github.com/Expensify/App/issues/7462) was occuring only on native platforms, longpressing on images worked fine on `mWeb`.
3. That means, we can remove `LongPressGestureHandler` from  [PressableWithSecondaryInteraction/index.js](https://github.com/Expensify/App/blob/main/src/components/PressableWithSecondaryInteraction/index.js) and add onLongPress listener to [Pressable](https://github.com/Expensify/App/blob/3faa6a77b076f6e6547340b93a2298042cf36266/src/components/PressableWithSecondaryInteraction/index.js#L70):
```javascript
onLongPress={this.callSecondaryInteractionWithMappedEvent}
```

https://user-images.githubusercontent.com/9059945/166672914-0de59ede-13f4-46db-97be-91b745f74f54.mp4



This resolves the current issue and doesn't break the [#7462](https://github.com/Expensify/App/issues/7462)

**But**
While testing this on other platforms i found a similar issue on iOS, which is similar to this, has different root cause, but still caused by LongPressGestureHandler. Pressing on plain link works fine on iOS, but then you have a message that contains text **and** link, regular context menu is displayed

https://user-images.githubusercontent.com/9059945/166672883-57751255-0649-41df-81b9-43ac39b65ea3.mov

But the cause is different from web, where both LongPressGestureHandler listeners are called, on iOS only the parent's listener get's called. However, i'm not sure this is not a fluke of dev mode/simulator, since I'm getting mixed results when testing this (sometimes it works as expected, sometimes it doesn't) and when running [snack](https://snack.expo.dev/rLAPULxbW) on iOS, i'm also getting random results. 

Can someone plese check if this issue present when runnig production build on physical iOS device?

**Comment by parasharrajat** [comment]
Thanks for the updated proposal @eVoloshchak. Seems to me the best proposal so far but let's discuss the iOS issue you mentioned first. Unfortunately, I don't have iOS. Let's see who can help us here to test it.

**Comment by parasharrajat** [comment]
I tested it on your snack but I also found the same issue. After some digging, I found out that iOS handle events differently, and all the handlers are triggered simultaneously when touches are overlapping. https://github.com/software-mansion/react-native-gesture-handler/issues/1930#issuecomment-1068920275. 

They suggest using https://docs.swmansion.com/react-native-gesture-handler/docs/gesture-composition/#simultaneous. I don't know how they can be used. 

Let me know if you find something.

**Comment by eVoloshchak** [comment]
> I found out that iOS handle events differently, and all the handlers are triggered simultaneously when touches are overlapping. [software-mansion/react-native-gesture-handler#1930 (comment)](https://github.com/software-mansion/react-native-gesture-handler/issues/1930#issuecomment-1068920275).
> 
> They suggest using https://docs.swmansion.com/react-native-gesture-handler/docs/gesture-composition/#simultaneous. I don't know how they can be used.

@parasharrajat , this is a great find actually, I was lookig for something like this, but never managed to find it.
But we can't use this in our scenario (at least I couldn't figure out a way to).
Besides, as it turns out, this is not iOS specific, this problem is also present on Android (the issue description should probably be updated). Same as with iOS, wrong context menu is displayed when we have a message with link and text.

https://user-images.githubusercontent.com/9059945/167257217-cbb1fb21-7723-4f77-bd3f-e996a0aaac8b.mp4

I've recreated out setup in [this snack](https://snack.expo.dev/O6FQ4pPUD), you can see, that no matter where you long press, only the parent handler registers the touch. And this is not a problem with nested gesture handlers either. Even if you delete the parent handler (const App, line 18), the handler in AnchorRenderer doesn't get called. It starts working if you wrap Text into View in AnchorRenderer, but i'm guessing that would break [#4911](https://github.com/Expensify/App/issues/4911).

But as this is the case with mWeb, we can look into why `LongPressHandler` was added in the first place.
As I mentioned previuosly, it was added to fix  https://github.com/Expensify/App/issues/7462.
So actually, we need `LongPressHandler` only for messages with attachments, not for all the messages. 
That means we can:
1. Add an `isAttachment` method to [ReportActionItem.js](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionItem.js)
```javascript
isAttachment() {
    const message = _.last(lodashGet(this.props.action, 'message', [{}]));
    const isAttachment = _.has(this.props.action, 'isAttachment')
        ? this.props.action.isAttachment
        : ReportUtils.isReportMessageAttachment(message);

    return isAttachment;
}
```
2. Then pass it to `PressableWithSecondaryInteraction` [here](https://github.com/Expensify/App/blob/724ccb46ca4b2f48f6052592cbacc7ffe9528c5c/src/pages/home/report/ReportActionItem.js#L154) 
```javascript
useLongPressHandler={this.isAttachment()}
```
3. And in [PressableWithSecondaryInteraction/index.native.js](https://github.com/Expensify/App/blob/main/src/components/PressableWithSecondaryInteraction/index.native.js) use `LongPressHandler` only for attachments
```javascript
const PressableWithSecondaryInteraction = (props) => {
    // Use Text node for inline mode to prevent content overflow.
    const Node = props.inline ? Text : Pressable;
    if (props.useLongPressHandler) {
        return (
            <LongPressGestureHandler
                onHandlerStateChange={(e) => {
                    if (e.nativeEvent.state !== State.ACTIVE) {
                        return;
                    }
                    e.preventDefault();
                    HapticFeedback.trigger();
                    props.onSecondaryInteraction(e);
                }}
            >
                <Node
                    ref={props.forwardedRef}
                    onPress={props.onPress}
                    onPressIn={props.onPressIn}
                    onPressOut={props.onPressOut}
                // eslint-disable-next-line react/jsx-props-no-spreading
                    {...(_.omit(props, 'onLongPress'))}
                >
                    {props.children}
                </Node>
            </LongPressGestureHandler>
        );
    }

    return (
        <Node
            ref={props.forwardedRef}
            onPress={props.onPress}
            onPressIn={props.onPressIn}
            onPressOut={props.onPressOut}
            onLongPress={(e) => {
                e.preventDefault();
                HapticFeedback.trigger();
                props.onSecondaryInteraction(e);
            }}
        // eslint-disable-next-line react/jsx-props-no-spreading
            {...props}
        >
            {props.children}
        </Node>
    )
};
```

**Android:** 

https://user-images.githubusercontent.com/9059945/167259600-8c55bf45-df90-4c3f-a40d-a9a636114c8d.mp4


**iOS:**

https://user-images.githubusercontent.com/9059945/167259617-e988c14f-a35d-4901-bbe3-26584e4d8fe2.mov

**Comment by parasharrajat** [comment]
Looking now.

**Comment by parasharrajat** [comment]
@eVoloshchak this proposals somewhat match with the approach from the previous [proposal](https://github.com/Expensify/App/issues/8311#issuecomment-1101199192).  This is doable but I don't think it would be much helpful. It is limited to this scenario only. If we are about to make any change it can break.

What if we manage the nested handlers. I have a few crazy ideas. 
-  Utilize this https://docs.swmansion.com/react-native-gesture-handler/docs/2.0.0/gesture-handlers/basics/interactions#simultaneous-recognition. 
- Or create event bubbling. 

 - We create a component that creates a `LongPressGestureHandler` around the children.
 - This component will save the ref with a nesting level in the Context.
 - Context value can be `Array<{ref: Ref, level: Number}>`. we will push each instance on the context value.
 - Now as soon as any handler's gesture is activated, we bubble the event to each handler. It can be done via updating the contextValue and setting active true one by one based on level `{ref: Ref, level: Number, active: Boolean}`.
 - Updating the Context will trigger an update on each component from step one. 
 - We can also cancel the event at any level. We just need to prevent bubbling. 
 
 What do you say? This will allow us to use this anywhere in the app.

**Comment by eVoloshchak** [comment]
> What do you say? This will allow us to use this anywhere in the app.

Sounds fun :) I'll try to implement this, will come back with results in ~3 days

**Comment by parasharrajat** [comment]
I am not saying that it is the way but thought of sharing.

**Comment by eVoloshchak** [comment]
@parasharrajat , just realised this is not gonna work, since the issue isn't with nested gesture handlers.
Take a look at [this](https://snack.expo.dev/0j4-grHwE) snack, it has only one `LongPressGestureHandler`, but still doesn't work.
However, if we take the same snack, but wrap our url `Text` with `View`, it starts working - [snack](https://snack.expo.dev/Ytx0AjUFo).
I'll try to look into **why** wrapping `View` with `LongPressGestureHandler`, but wrapping `Text` doesn't

**Comment by marcochavezf** [comment]
Not overdue, waiting for proposals.

**Comment by arielgreen** [comment]
> I will open a slack discussion to discuss existing proposals. But this is still open for better proposals.

@parasharrajat are we still waiting?

**Comment by parasharrajat** [comment]
Yeah, I will open the discussion soon. Need to summarize the discussion. But I open to proposals.

**Comment by parasharrajat** [comment]
I will open the discussion on slack shortly...

**Comment by marcochavezf** [comment]
> I will open the discussion on slack shortly...

Hi @parasharrajat! :) any update on this?

**Comment by parasharrajat** [comment]
Raised https://expensify.slack.com/archives/C01GTK53T8Q/p1656349294099259.

**Comment by marcochavezf** [comment]
Responding in the thread

**Comment by ahmdshrif** [comment]
### Proposal 

it's a small problem and does not need all these changes. 

#### Problem 
the 2 events trigger ( long-press on link, long-press on the comment) 
the link event first and then the comment event. 
and both contextMenu is show at the same time but the second one will override the first one 

![Screen Shot 2022-07-14 at 4 16 14 PM](https://user-images.githubusercontent.com/21364901/179003804-f8119332-324b-4860-811e-0754b7c6d4a5.png)

#### Solution : 

just prevent event default on the first long press  (that will be the child )
we already do this on native and that is why it works on native. 

https://github.com/Expensify/App/blob/ce899227ca6d3823e2865a5e194b429f1263db30/src/components/PressableWithSecondaryInteraction/index.native.js#L24

#### changes : 
on 

https://github.com/Expensify/App/blob/ce899227ca6d3823e2865a5e194b429f1263db30/src/components/PressableWithSecondaryInteraction/index.js#L38

```diff
    callSecondaryInteractionWithMappedEvent(e) {
        if ((e.nativeEvent.state !== State.ACTIVE) || hasHoverSupport()) {
            return;
        }
+        e.preventDefault();

     ...
       }
 ```
 
#### Result : 

https://user-images.githubusercontent.com/21364901/179004902-7fe25915-c68f-491c-80d9-04fb93369ea1.mov


cc : @parasharrajat

**Comment by marcochavezf** [comment]
Not overdue, we have a [new proposal ](https://github.com/Expensify/App/issues/8311#issuecomment-1184512351) to be reviewed.

**Comment by parasharrajat** [comment]
Does not work on Native and this issue also exists on native. 


https://user-images.githubusercontent.com/24370807/179632298-2d72fca9-4eee-4055-9fda-599a3d11e344.mp4

**Comment by ahmdshrif** [comment]
yes, this [proposal](https://github.com/Expensify/App/issues/8311#issuecomment-1184512351) for web only because the issue says it's mWeb.

>Where is this issue occurring?
>Mobile Web

on native, it's another root cause but I think it is worth bouns. 

any way. 

### proposal  ( for native) 
the root cause is the child of `LongPressGestureHandler`  is a text .
and it need a container for it so we can use view or pressable and that will work 


as we use text component for aligning we will have issues with aligning but we can solve it by cauterizing comment rendering.

#### changes :
https://github.com/Expensify/App/blob/6369e8e3136b95e03205c5bde751028734600071/src/components/PressableWithSecondaryInteraction/index.native.js#L29
replace node with pressable 

```ts
        <Pressable
            ref={props.forwardedRef}
            onPress={props.onPress}
            onPressIn={props.onPressIn}
            onPressOut={props.onPressOut}
            // eslint-disable-next-line react/jsx-props-no-spreading
            {...(_.omit(props, 'onLongPress'))}
        >
            {props.children}
        </Pressable>
```

this change will make links work fine on native but the text algin will need more work.

I fix it by customizing the comment rendered on HTML engine 
and wrap comment children on view with alginContent center. 

```ts
const CommentRender = props => (
    <View style={{alignContent: 'center', flexDirection: 'row'}}>
        { _.map(props.tnode.children, node => (<TNodeChildrenRenderer tnode={node} />))}
    </View>
);
``` 

also, there is some additional point that makes link better clicking by increase the [maxDix](https://docs.swmansion.com/react-native-gesture-handler/docs/gesture-handlers/api/longpress-gh/#maxdist) to 20 (improve only) 
```ts
    <LongPressGestureHandler
...
        maxDist={20}
    >
...
```

full changes are here  https://github.com/ahmdshrif/App/compare/main...ahmdshrif:App:fix/link-longpress ( i include support for multiLine comment also [here](https://github.com/ahmdshrif/App/commit/44307b0dd0eadecbf78481fb49938dda29cf0721)  )

cc: @parasharrajat  @marcochavezf.

**Comment by parasharrajat** [comment]
I agree issue details are incomplete but many times it is case that only a single platform is mentioned but all issues need the be fixed on all platforms.

As the same issue occur on native too, it needs the be fixed as well.


At this time, scope of the issue is not extended so I don't see a bonus but it can be.

@kavimuru could you please add Android and iOS platform as well?

**Comment by mdneyazahmad** [comment]
On my android device production app (v1.1.79-17), it sometime works and sometime does not work.

cc: @parasharrajat @ahmdshrif

**Comment by mdneyazahmad** [comment]
On my android device production app (v1.1.79-17), it sometime works and sometime does not work.

cc: @parasharrajat @ahmdshrif

**Comment by parasharrajat** [comment]
@ahmdshrif Could you please only suggest changes that are necessary to fix this issue in your proposal?

> Change LongPressGestureHandler to Pressable.

Also, if you look back the history, you made this change to allow propagation of long press for Attachments. (Don't remember exactly but you can find it) so that should not break. 

I do not understand the purpose of CommentRender. What is going on here and why? I don't HTML manipulation is necessary for this issue at all.

```
   const nodes = props.tnode.children[0].children;
    _.map(nodes, (node) => {
        if (node.tagName !== 'br') {
            currentLine.push(node);
            return;
        }
        lines.push(currentLine);
        currentLine = [];
    });
    if (currentLine.length > 0) {
        lines.push(currentLine);
    }
```

**Comment by ahmdshrif** [comment]
@parasharrajat  yes I remember my change and I don't break it. 

the simple proposal for test is this change in PressableWithSecondaryInteraction/index.native.js
```ts
  <LongPressGestureHandler
        onHandlerStateChange={(e) => {
            if (e.nativeEvent.state !== State.ACTIVE) {
                return;
            }
            e.preventDefault();
            HapticFeedback.trigger();
            props.onSecondaryInteraction(e);
        }}
        maxDist={20}
        minDurationMs={props.inline ? 400 : 500}
    >
        <Pressable
            ref={props.forwardedRef}
            onPress={props.onPress}
            onPressIn={props.onPressIn}
            onPressOut={props.onPressOut}
            // eslint-disable-next-line react/jsx-props-no-spreading
            {...(_.omit(props, 'onLongPress'))}
        >
            {props.children}
        </Pressable>
    </LongPressGestureHandler>
```
and  in 
PressableWithSecondaryInteraction/index..js
```ts
  callSecondaryInteractionWithMappedEvent(e) {
        if ((e.nativeEvent.state !== State.ACTIVE) || hasHoverSupport()) {
            return;
        }
        e.preventDefault();

     ...
       }
```

this will break the text alignment (as you see in screenshots)  so other code to fix this alignment.   I can explain more about how it work but for now, let's get confirmation about the main fix for long-press. 

![Screen Shot 2022-07-20 at 4 37 33 PM](https://user-images.githubusercontent.com/21364901/180009857-dda8e093-5df2-4ef3-b905-024fab49da59.png)

**Comment by arielgreen** [comment]
Reposted since the job closed https://www.upwork.com/jobs/~0106b378e281080b23

**Comment by b1tjoy** [comment]
### PROPOSAL
### Problem
We use LongPressGestureHandler to replace Pressable in PR https://github.com/Expensify/App/pull/8018, and this PR fix https://github.com/Expensify/App/issues/7462. In that issue, long press on image doesn't open the context menu, but act like short press and open image thumbnail. If we fix issue 7462 with a solution which does not replace Pressable with LongPressGestureHandler, our current issue will be fixed too.

The root cause of issue 7462 is the inner TouchableWithoutFocus's onPress handler conflicts with outer Pressable onLongPress handler, they both use react-native Gesture Responder System, and they can not coordinate with each other.

### Solution
If we use RNGH's TapGestureHandler to replace TouchableOpacity, and revert PR https://github.com/Expensify/App/pull/8018, both https://github.com/Expensify/App/issues/7462 and https://github.com/Expensify/App/issues/8311 will be fixed. Because RNGH does not use Gesture Responder System of react-native, press and long press can work with each other.

Here is the solution:

1. Revert changes made by PR https://github.com/Expensify/App/pull/8018
2. Use TapGestureHandler in TouchableWithoutFocus.js as the following code

Replace
https://github.com/Expensify/App/blob/03b953404d3fc17d03fc5635322f4e12c25afb7b/src/components/TouchableWithoutFocus.js#L39-L41
With
```JSX
<TapGestureHandler
  onHandlerStateChange={(event) => {
    if (event.nativeEvent.state === State.ACTIVE) {
      this.touchableRef.blur();
      this.props.onPress();
    }
  }}
>
  <View ref={el => this.touchableRef = el} style={this.props.styles}>
    {this.props.children}
  </View>
</TapGestureHandler>
```

Since PR https://github.com/Expensify/App/pull/8018 was merged in 1.1.43-0, I use tag 1.1.42-6 in our repo above to verify the solution. I will post a complete proposal based on main repo later.

### Screenshot
The screenshot was also taken with tag 1.1.42-6 to demonstrate the solution can fix both issues, will update with new screenshot later.

https://user-images.githubusercontent.com/103875612/182896277-acdd24ac-a956-452d-9558-a6ec26876776.mp4

**Comment by b1tjoy** [comment]
### UPDATE
This is a update of proposal https://github.com/Expensify/App/issues/8311#issuecomment-1205476421, change the solution code based on the latest main branch.

### PROPOSAL

### Problem

We use `LongPressGestureHandler` to replace `Pressable` in PR https://github.com/Expensify/App/pull/8018, and this PR fix https://github.com/Expensify/App/issues/7462. In that issue, long press on image doesn't open the context menu, but act like short press and open image thumbnail. If we fix issue 7462 with a solution which does not replace `Pressable` with `LongPressGestureHandler`, our current issue will be fixed too.

The root cause of issue 7462 is the inner PressableWithoutFocus's onPress handler conflicts with outer Pressable onLongPress handler, they both use react-native Gesture Responder System, and they can not coordinate with each other.

### Solution

If we revert PR https://github.com/Expensify/App/pull/8018, and use RNGH's `TapGestureHandler` to replace `Pressable`, both https://github.com/Expensify/App/issues/7462 and https://github.com/Expensify/App/issues/8311 will be fixed. Because RNGH does not use Gesture Responder System of react-native, inner press handler and outer long press handler can work with each other.

1. Revert PR https://github.com/Expensify/App/pull/8018

https://github.com/Expensify/App/blob/97eb8bb0019068b484035a17a64e4ecdeb147092/src/components/PressableWithSecondaryInteraction/index.js
```JSX
diff --git a/src/components/PressableWithSecondaryInteraction/index.js b/src/components/PressableWithSecondaryInteraction/index.js
index a6b21a59e..78168a9f4 100644
--- a/src/components/PressableWithSecondaryInteraction/index.js
+++ b/src/components/PressableWithSecondaryInteraction/index.js
@@ -1,7 +1,6 @@
 import _ from 'underscore';
 import React, {Component} from 'react';
 import {Pressable} from 'react-native';
-import {LongPressGestureHandler, State} from 'react-native-gesture-handler';
 import * as pressableWithSecondaryInteractionPropTypes from './pressableWithSecondaryInteractionPropTypes';
 import styles from '../../styles/styles';
 import hasHoverSupport from '../../libs/hasHoverSupport';
@@ -12,7 +11,6 @@ import hasHoverSupport from '../../libs/hasHoverSupport';
 class PressableWithSecondaryInteraction extends Component {
     constructor(props) {
         super(props);
-        this.callSecondaryInteractionWithMappedEvent = this.callSecondaryInteractionWithMappedEvent.bind(this);
         this.executeSecondaryInteractionOnContextMenu = this.executeSecondaryInteractionOnContextMenu.bind(this);
     }
 
@@ -27,27 +25,6 @@ class PressableWithSecondaryInteraction extends Component {
         this.pressableRef.removeEventListener('contextmenu', this.executeSecondaryInteractionOnContextMenu);
     }
 
-    /**
-     * @param {Object} e
-     */
-    callSecondaryInteractionWithMappedEvent(e) {
-        if ((e.nativeEvent.state !== State.ACTIVE) || hasHoverSupport()) {
-            return;
-        }
-
-        // Map gesture event to normal Responder event
-        const {
-            absoluteX, absoluteY, locationX, locationY,
-        } = e.nativeEvent;
-        const mapEvent = {
-            ...e,
-            nativeEvent: {
-                ...e.nativeEvent, pageX: absoluteX, pageY: absoluteY, x: locationX, y: locationY,
-            },
-        };
-        this.props.onSecondaryInteraction(mapEvent);
-    }
-
     /**
      * @param {contextmenu} e - A right-click MouseEvent.
      * https://developer.mozilla.org/en-US/docs/Web/API/Element/contextmenu_event
@@ -65,19 +42,23 @@ class PressableWithSecondaryInteraction extends Component {
 
         // On Web, Text does not support LongPress events thus manage inline mode with styling instead of using Text.
         return (
-            <LongPressGestureHandler onHandlerStateChange={this.callSecondaryInteractionWithMappedEvent}>
-                <Pressable
-                    style={this.props.inline && styles.dInline}
-                    onPressIn={this.props.onPressIn}
-                    onPressOut={this.props.onPressOut}
-                    onPress={this.props.onPress}
-                    ref={el => this.pressableRef = el}
-                    // eslint-disable-next-line react/jsx-props-no-spreading
-                    {...defaultPressableProps}
-                >
-                    {this.props.children}
-                </Pressable>
-            </LongPressGestureHandler>
+            <Pressable
+                style={this.props.inline && styles.dInline}
+                onPressIn={this.props.onPressIn}
+                onLongPress={(e) => {
+                    if (hasHoverSupport()) {
+                        return;
+                    }
+                    this.props.onSecondaryInteraction(e);
+                }}
+                onPressOut={this.props.onPressOut}
+                onPress={this.props.onPress}
+                ref={el => this.pressableRef = el}
+                // eslint-disable-next-line react/jsx-props-no-spreading
+                {...defaultPressableProps}
+            >
+                {this.props.children}
+            </Pressable>
         );
     }
 }
```

https://github.com/Expensify/App/blob/main/src/components/PressableWithSecondaryInteraction/index.native.js
```JSX
diff --git a/src/components/PressableWithSecondaryInteraction/index.native.js b/src/components/PressableWithSecondaryInteraction/index.native.js
index efb68e52e..8c6c145cc 100644
--- a/src/components/PressableWithSecondaryInteraction/index.native.js
+++ b/src/components/PressableWithSecondaryInteraction/index.native.js
@@ -1,7 +1,6 @@
 import _ from 'underscore';
 import React, {forwardRef} from 'react';
 import {Pressable} from 'react-native';
-import {LongPressGestureHandler, State} from 'react-native-gesture-handler';
 import * as pressableWithSecondaryInteractionPropTypes from './pressableWithSecondaryInteractionPropTypes';
 import Text from '../Text';
 import HapticFeedback from '../../libs/HapticFeedback';
@@ -16,28 +15,21 @@ const PressableWithSecondaryInteraction = (props) => {
     // Use Text node for inline mode to prevent content overflow.
     const Node = props.inline ? Text : Pressable;
     return (
-        <LongPressGestureHandler
-            onHandlerStateChange={(e) => {
-                if (e.nativeEvent.state !== State.ACTIVE) {
-                    return;
-                }
+        <Node
+            ref={props.forwardedRef}
+            onPress={props.onPress}
+            onPressIn={props.onPressIn}
+            onLongPress={(e) => {
                 e.preventDefault();
                 HapticFeedback.trigger();
                 props.onSecondaryInteraction(e);
             }}
+            onPressOut={props.onPressOut}
+        // eslint-disable-next-line react/jsx-props-no-spreading
+            {...(_.omit(props, 'onLongPress'))}
         >
-            <Node
-                ref={props.forwardedRef}
-                onPress={props.onPress}
-                onPressIn={props.onPressIn}
-                onPressOut={props.onPressOut}
-            // eslint-disable-next-line react/jsx-props-no-spreading
-                {...(_.omit(props, 'onLongPress'))}
-            >
-                {props.children}
-            </Node>
-        </LongPressGestureHandler>
-
+            {props.children}
+        </Node>
     );
 };
```

2. Use TapGestureHandler to handle press event in https://github.com/Expensify/App/blob/main/src/components/PressableWithoutFocus.js

```JSX
diff --git a/src/components/PressableWithoutFocus.js b/src/components/PressableWithoutFocus.js
index 35a9cce99..87cf6eb1c 100644
--- a/src/components/PressableWithoutFocus.js
+++ b/src/components/PressableWithoutFocus.js
@@ -1,6 +1,7 @@
 import React from 'react';
-import {Pressable} from 'react-native';
+import {View} from 'react-native';
 import PropTypes from 'prop-types';
+import {State, TapGestureHandler} from 'react-native-gesture-handler';
 
 const propTypes = {
     /** Element that should be clickable  */
@@ -33,16 +34,21 @@ class PressableWithoutFocus extends React.Component {
         this.pressAndBlur = this.pressAndBlur.bind(this);
     }
 
-    pressAndBlur() {
+    pressAndBlur(event) {
+        if (event.nativeEvent.state !== State.ACTIVE) {
+            return;
+        }
         this.pressableRef.blur();
         this.props.onPress();
     }
 
     render() {
         return (
-            <Pressable onPress={this.pressAndBlur} ref={el => this.pressableRef = el} style={this.props.styles}>
-                {this.props.children}
-            </Pressable>
+            <TapGestureHandler onHandlerStateChange={this.pressAndBlur}>
+                <View ref={el => this.pressableRef = el} style={this.props.styles}>
+                    {this.props.children}
+                </View>
+            </TapGestureHandler>
         );
     }
 }
```

### Screenshot for issue 7462
Long press on image opens context menu, press on image opens attachment model.

iOS native screenshot

https://user-images.githubusercontent.com/103875612/183334489-eeb890cc-3c37-4544-8532-ec4451e53323.mp4



Android native screenshot

https://user-images.githubusercontent.com/103875612/183339358-af324fcf-fbb3-4a3d-ad8d-24d12bc52a25.mp4


mWeb screenshot


### Screenshot for issue 8311
Long press on email/URL opens Copy to clipboard context menu, long press on message opens full context menu.

iOS native screenshot

https://user-images.githubusercontent.com/103875612/183334520-f9d26a97-e408-4603-a55a-9d66e9e771f2.mp4



Android native screenshot

https://user-images.githubusercontent.com/103875612/183339381-efa8f4b9-63d1-4834-987c-2f32b4ab9bce.mp4


mWeb screenshot

**Comment by ahmdshrif** [comment]
@parasharrajat  we already have this alignment issue and if we can avoid it now it will appear in future changes. 

I know you refactor this file in your PR  but take a look at this comment. 
https://github.com/Expensify/App/blob/97eb8bb0019068b484035a17a64e4ecdeb147092/src/components/HTMLEngineProvider/HTMLRenderers/AnchorRenderer.js#L58

the issue is htmlRender map comment  container to text and text can be use to align the text if all child is text only  not text inside any container. 
for example :
this will be aligned 
```ts
<Text >
        <Text >text 1 </Text>
        <Text >text 2 </Text>
</Text>
```

but this will not be aligned in android
```ts
<Text >
       <View>
        <Text>text 1 </Text>
      </View>
      <Text>text 2</Text>
</Text>
```
check this [snack](https://snack.expo.dev/@as1171111/15f1cd) on android 

and the second example happened with us in this case and on the text link case. 

I think we need to get out of this issue and have some flexibility to add wherever we need on the comment.
we just need to change the container of the `comment text` to a `view` component not  `Text` with some alignment style .  as I post before. 

and HTMLRender manages this part so I don't think if we add some customize to the container and keep HTMLRender to managing the child is a bad change.

**Comment by ahmdshrif** [comment]
but this issue appears only because we customize some tags. so I don't think we need to change the lib structure to meet our customize tags .maybe it will conflict with someone else customization.

**Comment by arielgreen** [comment]
not overdue

**Comment by songlang1994** [comment]
### Proposal

`LongPressGestureHandler` was introduced in PR https://github.com/Expensify/App/pull/8018 which tried to resolve the nested `Pressable` issue.

```jsx
import {Pressable} from 'react-native'

<Pressable onPress={} onLongPress={}>  // <-- onPress & onLongPress won't be triggered if user tap the inner Pressable
    <Pressable onPress={}></Pressable>
</Pressable>
```

We can wrap `Pressable` to make `onPress` and `onLongPress` events bubble up, just like in a normal browser environment. Then replace all `Pressable` (from react-native) with the new wrapped version and everything should be work as expected.

Verified on mWeb and Android.

```jsx
// File: src/components/Pressable.js
import _ from 'underscore';
import React, {useContext, useImperativeHandle, useRef} from 'react';
import {Pressable as RNPressable} from 'react-native';

const Context = React.createContext(null);

// Accepts the same parameters as react-native `Pressable` but `onPress` & `onLongPress` will bubble up
// to the root `Pressable` unless `event.stopPropagation()` is called.
function Pressable(props) {
    const defaultPressableProps = _.omit(props, ['onPress', 'onLongPress', 'children']);

    const parentRef = useContext(Context);
    const ref = useRef(null);

    const onPressProxy = (event) => {
        if (props.onPress) {
            props.onPress(event);
        }

        if (event.isPropagationStopped()) {
            return;
        }
        setTimeout(() => {
            if (!(parentRef && parentRef.current)) {
                return;
            }
            parentRef.current.onPress(event);
        }, 0);
    };

    const onLongPressProxy = (event) => {
        if (props.onLongPress) {
            props.onLongPress(event);
        }

        if (event.isPropagationStopped()) {
            return;
        }
        setTimeout(() => {
            if (!(parentRef && parentRef.current)) {
                return;
            }
            parentRef.current.onLongPress(event);
        }, 0);
    };

    useImperativeHandle(ref, () => ({
        onPress: onPressProxy,
        onLongPress: onLongPressProxy,
    }));

    return (
        <RNPressable
            ref={props.forwardedRef}
            // eslint-disable-next-line react/jsx-props-no-spreading
            {...defaultPressableProps}
            onPress={onPressProxy}
            onLongPress={onLongPressProxy}
        >
            <Context.Provider value={ref}>
                {props.children}
            </Context.Provider>
        </RNPressable>
    );
}

Pressable.propTypes = RNPressable.propTypes;
export default React.forwardRef((props, ref) => (
    // eslint-disable-next-line react/jsx-props-no-spreading
    <Pressable {...props} forwardedRef={ref} />
));

```

**Comment by songlang1994** [comment]
> Can't we just do it for PressableWithSecondaryInteraction? 

At least we need to do it for `PressableWithSecondaryInteraction` and `PressableWithoutFocus` (used by `ImageRenderer`). Another way is creating a separated `PressableWithoutFocus2` for `ImageRender` If you're worried about directly modifing `PressableWithout` too much scope.

The above proposal https://github.com/Expensify/App/issues/8311#issuecomment-1207653523 says to use a different gesture handler component to handle the nested case. I don't think it's general enough. It can only handle 2 level nesting and the inner `Pressable` can not stop the event propagation.

```
<TapGestureHandler> <-- react-native-gesture-handler
  <Pressable></Pressable> <-- react-native
</TapGestureHandler>
```

### Update
If you don't want to change the default behavior of react-native `Pressable`, I recommend renaming the new component to `NestablePressable` and using it only in necessary components.

**Comment by marcochavezf** [comment]
not overdue, hi @parasharrajat what do you think about the `NestablePressable` strategy that @songlang1994 mentioned in the last comment? Seems that a different nested `Pressable` is a good idea to prevent events bubble up.

**Comment by songlang1994** [comment]
@parasharrajat The Context API is required but the Hooks API is not. We can use a class component to achieve the same purpose. Verified on mWeb and Android.

```js
// File: src/components/NestablePressable.js

import _ from 'underscore';
import React from 'react';
import {Pressable as RNPressable} from 'react-native';

const Context = React.createContext(null);

class NestablePressable extends React.Component {
    constructor(props) {
        super(props);
        this.onPress = this.onPress.bind(this);
        this.onLongPress = this.onLongPress.bind(this);
    }

    onPress(event) {
        this.callEventHandler('onPress', event);
    }

    onLongPress(event) {
        this.callEventHandler('onLongPress', event);
    }

    getAncestorComponent() {
        return this.context;
    }

    callEventHandler(name, event) {
        if (this.props[name]) {
            this.props[name](event);
        }

        if (event.isPropagationStopped()) {
            return;
        }

        setTimeout(() => {
            if (!this.getAncestorComponent()) {
                return;
            }
            this.getAncestorComponent()[name](event);
        }, 0);
    }

    render() {
        const defaultPressableProps = _.omit(this.props, ['onPress', 'onLongPress', 'children']);

        return (
            <RNPressable
                ref={this.props.forwardedRef}
                // eslint-disable-next-line react/jsx-props-no-spreading
                {...defaultPressableProps}
                onPress={this.onPress}
                onLongPress={this.onLongPress}
            >
                <Context.Provider value={this}>
                    {this.props.children}
                </Context.Provider>
            </RNPressable>
        );
    }
}

NestablePressable.contextType = Context;
export default React.forwardRef((props, ref) => (
    // eslint-disable-next-line react/jsx-props-no-spreading
    <NestablePressable {...props} forwardedRef={ref} />
));
```

**Comment by arielgreen** [comment]
@parasharrajat do you have an update?

**Comment by songlang1994** [comment]
I'm in the same situation as above. Could you please invite me too?

Email: songlang1994@outlook.com

**Comment by arielgreen** [comment]
@songlang1994 @b1tjoy please link your Upwork profile as well. You can access it by clicking your avatar in the upper right corner then clicking your name in the upper right corner, it'll look like: https://www.upwork.com/freelancers/%5C%5C%5C%5C%5C%5C%5C%5C~019c6cd28a0575c222

**Comment by arielgreen** [comment]
@b1tjoy you should be added shortly

**Comment by songlang1994** [comment]
@arielgreen Upwork profile: https://www.upwork.com/freelancers/~018c1e5b7b47a9c2c0

**Comment by arielgreen** [comment]
done

**Comment by songlang1994** [comment]
@parasharrajat I read the discussion that you were concerned about using `setTimeout`, so I refactored `NestablePressable` to invoke all event handlers in the gesture handler component (a.k.a the inner component). This avoids using `setTimeout` and make the code cleaner.

```js
// File: src/components/NestablePressable.js

import _ from 'underscore';
import React from 'react';
import {Pressable as RNPressable} from 'react-native';

const Context = React.createContext(null);

class NestablePressable extends React.Component {
    constructor(props) {
        super(props);
        this.dispatchPressEvent = this.dispatchPressEvent.bind(this);
        this.dispatchLongPressEvent = this.dispatchLongPressEvent.bind(this);
    }

    getAncestorComponent() {
        return this.context;
    }

    dispatchPressEvent(event) {
        this.callEventHandlers('onPress', event);
    }

    dispatchLongPressEvent(event) {
        this.callEventHandlers('onLongPress', event);
    }

    callEventHandlers(name, event) {
        // eslint-disable-next-line consistent-this
        let target = this;

        while (target) {
            try {
                const handler = target.props[name];
                if (handler) {
                    handler(event);
                }
            } catch (err) {
                console.error(err);
            }
            if (event.isPropagationStopped()) {
                break;
            }
            target = target.getAncestorComponent();
        }
    }

    render() {
        const defaultPressableProps = _.omit(this.props, ['onPress', 'onLongPress', 'children']);

        return (
            <RNPressable
                ref={this.props.forwardedRef}
                // eslint-disable-next-line react/jsx-props-no-spreading
                {...defaultPressableProps}
                onPress={this.dispatchPressEvent}
                onLongPress={this.dispatchLongPressEvent}
            >
                <Context.Provider value={this}>
                    {this.props.children}
                </Context.Provider>
            </RNPressable>
        );
    }
}

NestablePressable.propTypes = RNPressable.propTypes;
NestablePressable.contextType = Context;
export default React.forwardRef((props, ref) => (
    // eslint-disable-next-line react/jsx-props-no-spreading
    <NestablePressable {...props} forwardedRef={ref} />
));
```

**Comment by marcochavezf** [comment]
Commenting on the [slack discussion](https://expensify.slack.com/archives/C01GTK53T8Q/p1663015463167629?thread_ts=1662392381.977999&cid=C01GTK53T8Q).

**Comment by marcochavezf** [comment]
Not overdue, discussing on slack an alternative solution

**Comment by songlang1994** [comment]
### Updated Proposal

According to the discussion in previous several days, we decided to change the popover menu behavior to *“App should also show generic options at the same time when the user aux-click/long-press on a link/email/image element”*. This proposal is based on my previous comment (https://github.com/Expensify/App/issues/8311#issuecomment-1239023718) to implements the new behavior.

#### Video demos
<details>
  <summary>Click to watch the videos.</summary>

https://user-images.githubusercontent.com/19236594/190606528-9079c9f7-5193-4354-96af-1d0e9e88a69c.mp4

https://user-images.githubusercontent.com/19236594/190607430-65bcfb83-9a40-4377-8e07-ae76b22c4363.mp4

<p>
</details>

#### Steps

1. Create a new file `src/pages/home/report/ReportActionItemContext.js`

```js
import {createContext} from 'react';
export default createContext(null);
```

2. In `src/pages/home/report/ReportActionItem.js`, refine `showPopover` and pass it through `ReportActionItemContext`. All descendant elements need to show the popover through `ReportActionItem.showPopover`.

```js
import ReportActionItemContext from './ReportActionItemContext';

class ReportActionItem extends Component {
    //...

    /*
    showPopover(event) {
        // Block menu on the message being Edited
        if (this.props.draftMessage) {
            return;
        }
        const selection = SelectionScraper.getCurrentSelection();
        ReportActionContextMenu.showContextMenu(
            ContextMenuActions.CONTEXT_MENU_TYPES.REPORT_ACTION,
            event,
            selection,
            this.popoverAnchor,
            this.props.reportID,
            this.props.action,
            this.props.draftMessage,
            this.checkIfContextMenuActive,
            this.checkIfContextMenuActive,
        );
    } */

    showPopover(event, type, selection, anchor) {
        // Block menu on the message being Edited
        if (this.props.draftMessage) {
            return;
        }
        ReportActionContextMenu.showContextMenu(
            type || ContextMenuActions.CONTEXT_MENU_TYPES.REPORT_ACTION,
            event,
            selection || SelectionScraper.getCurrentSelection(),
            anchor || this.popoverAnchor,
            this.props.reportID,
            this.props.action,
            this.props.draftMessage,
            this.checkIfContextMenuActive,
            this.checkIfContextMenuActive,
        );
    }

    // ...
    render() {
        return (
            <PressableWithSecondaryInteraction>
                <ReportActionItemContext.Provider value={{showPopover: this.showPopover}}>
                    // ...
                </ReportActionItemContext.Provider>
            </PressableWithSecondaryInteraction>
        )
    }
}
```

3. In `src/components/AnchorForCommentsOnly/BaseAnchorForCommentsOnly.js`, consume `showPopover` from the context.

```js
import ReportActionItemContext from '../../pages/home/report/ReportActionItemContext';

// In order to reduce the sample code I used the hook API. It can be completely replaced by class component.
const BaseAnchorForCommentsOnly = (props) => {
    // ...
    const context = useContext(ReportActionItemContext);

    return (
        <PressableWithSecondaryInteraction
            onSecondaryInteraction={(event) => {
                if (!context) {
                    return;
                }
                context.showPopover(
                    event,
                    Str.isValidEmail(props.displayName) ? ContextMenuActions.CONTEXT_MENU_TYPES.EMAIL : ContextMenuActions.CONTEXT_MENU_TYPES.LINK,
                    props.href,
                    lodashGet(linkRef, 'current'),
                );
            }}
        >
            // ...
        </PressableWithSecondaryInteraction>
    );
};
```

4. In `src/pages/home/report/ContextMenu/ContextMenuActions.js`, change `shouldShow` property.
```diff
--- a/src/pages/home/report/ContextMenu/ContextMenuActions.js
+++ b/src/pages/home/report/ContextMenu/ContextMenuActions.js
@@ -88,8 +88,7 @@ export default [
         icon: Expensicons.Clipboard,
         successTextTranslateKey: 'reportActionContextMenu.copied',
         successIcon: Expensicons.Checkmark,
-        shouldShow: (type, reportAction) => (type === CONTEXT_MENU_TYPES.REPORT_ACTION
-            && reportAction.actionName !== CONST.REPORT.ACTIONS.TYPE.IOU
+        shouldShow: (type, reportAction) => (reportAction.actionName !== CONST.REPORT.ACTIONS.TYPE.IOU
             && !ReportUtils.isReportMessageAttachment(_.last(lodashGet(reportAction, ['message'], [{}])))),
 
         // If return value is true, we switch the `text` and `icon` on
@@ -132,7 +131,7 @@ export default [
 
             // Only hide the copylink menu item when context menu is opened over img element.
             const isAttachmentTarget = lodashGet(menuTarget, 'tagName') === 'IMG' && isAttachment;
-            return Permissions.canUseCommentLinking(betas) && type === CONTEXT_MENU_TYPES.REPORT_ACTION && !isAttachmentTarget;
+            return Permissions.canUseCommentLinking(betas) && !isAttachmentTarget;
         },
         onPress: (closePopover, {reportAction, reportID}) => {
             Environment.getEnvironmentURL()
@@ -149,7 +148,7 @@ export default [
         textTranslateKey: 'reportActionContextMenu.markAsUnread',
         icon: Expensicons.Mail,
         successIcon: Expensicons.Checkmark,
-        shouldShow: type => type === CONTEXT_MENU_TYPES.REPORT_ACTION,
+        shouldShow: () => true,
         onPress: (closePopover, {reportAction, reportID}) => {
             Report.markCommentAsUnread(reportID, reportAction.sequenceNumber);
             if (closePopover) {
@@ -162,9 +161,7 @@ export default [
     {
         textTranslateKey: 'reportActionContextMenu.editComment',
         icon: Expensicons.Pencil,
-        shouldShow: (type, reportAction) => (
-            type === CONTEXT_MENU_TYPES.REPORT_ACTION && ReportUtils.canEditReportAction(reportAction)
-        ),
+        shouldShow: (type, reportAction) => ReportUtils.canEditReportAction(reportAction),
         onPress: (closePopover, {reportID, reportAction, draftMessage}) => {
             const editAction = () => Report.saveReportActionDraft(
                 reportID,
@@ -186,8 +183,7 @@ export default [
     {
         textTranslateKey: 'reportActionContextMenu.deleteComment',
         icon: Expensicons.Trashcan,
-        shouldShow: (type, reportAction) => type === CONTEXT_MENU_TYPES.REPORT_ACTION
-            && ReportUtils.canDeleteReportAction(reportAction),
+        shouldShow: (type, reportAction) => ReportUtils.canDeleteReportAction(reportAction),
         onPress: (closePopover, {reportID, reportAction}) => {
             if (closePopover) {
                 // Hide popover, then call showDeleteConfirmModal
```

**Comment by arielgreen** [comment]
Moved to weekly, @parasharrajat please provide an update

**Comment by parasharrajat** [comment]
Need to open a new discussion for a single context menu but now I feel like that might be a new feature unrelated to this issue. There are other issues at hand and we might not need the single context menu for a long time if we fix this issue.

Action item for me
1. Open a new discussion for the above.
2. Rereview the previous discussion and old proposals and decide if we can move forward with any of those.

**Comment by arielgreen** [comment]
@parasharrajat

**Comment by arielgreen** [comment]
@parasharrajat update please?

**Comment by parasharrajat** [comment]
Will be back on it on Friday/Saturday.

**Comment by arielgreen** [comment]
@parasharrajat what do we need to do to push this forward? We have a proposal that's been waiting two weeks for review.

**Comment by parasharrajat** [comment]
I will take it on by tomorrow and if possible today. All proposals are reviewed and we just need to make a final decision on which approach to take. The last time, I opened a discussion on slack there was no clear conclusion after waiting for days. So I will make a decision best of my judgment. Maybe it requires another discussion. Please wait till tomorrow.

**Comment by arielgreen** [comment]
Awaiting decision, but possibly delayed due to Diwali this week.

**Comment by arielgreen** [comment]
Waiting

**Comment by parasharrajat** [comment]
Discussing now.

**Comment by parasharrajat** [comment]
@b1tjoy Which one of your proposals is valid? Can you summarize all of your proposals? Add pointers to tell us which proposal is solving what problem.

Is [this](https://github.com/Expensify/App/issues/8311#issuecomment-1205476421) not working correctly?

**Comment by bernhardoj** [comment]
In case it still accepting a proposal, I would like to share mine.

### **Proposal**
To prevent the event bubbling to the parent handler, we can put `e.preventDefault()` inside `callSecondaryInteractionWithMappedEvent()` as below:

**_PressableWithSecondaryInteraction (index.js)_**
```diff
callSecondaryInteractionWithMappedEvent(e) {
        if ((e.nativeEvent.state !== State.ACTIVE) || hasHoverSupport()) {
            return;
        }
++        e.preventDefault();

        // Map gesture event to normal Responder event
        const {
            absoluteX, absoluteY, locationX, locationY,
        } = e.nativeEvent;
        const mapEvent = {
            ...e,
            nativeEvent: {
                ...e.nativeEvent, pageX: absoluteX, pageY: absoluteY, x: locationX, y: locationY,
            },
        };
        this.props.onSecondaryInteraction(mapEvent);
    }
```

We can take a look at the current native version of it which also use `e.preventDefault()`.

https://github.com/Expensify/App/blob/50e8901a3ddfbd298af485950b2d9a0cf8e12b7d/src/components/PressableWithSecondaryInteraction/index.native.js#L20-L28

**Result**

https://user-images.githubusercontent.com/50919443/199038054-1158959f-4d4b-4dba-b833-f244ccd6e602.mp4

**Comment by parasharrajat** [comment]
@bernhardoj You are free to post proposals if they are different from older ones.

**Comment by ahmdshrif** [comment]
@bernhardoj  this does not work on native. only work on web. I proposed it before https://github.com/Expensify/App/issues/8311?notification_referrer_id=NT_kwDOAUYApbMzMzI3MzkwNzE3OjIxMzY0OTAx#issuecomment-1184512351

**Comment by bernhardoj** [comment]
I see. I didn't see all the hidden discussions. My bad.

**Comment by parasharrajat** [comment]
Waiting on https://github.com/Expensify/App/issues/8311#issuecomment-1297168551.

**Comment by parasharrajat** [comment]
@b1tjoy both links are referring to same comment.

**Comment by parasharrajat** [comment]
Ok. Thanks. I will take that as your final proposal.

**Comment by marcochavezf** [comment]
> @parasharrajat You can review this one [#8311 (comment)](https://github.com/Expensify/App/issues/8311#issuecomment-1270322261) and ignore the others.

My two cents: I like this proposal because seems @b1tjoy already explored the edge cases in different platforms. Also, I'm in favor of keeping the current behavior of showing one pop-up for the email/URLs and another for the general case, it will keep things simple when there's a comment with multiple email/URL values. We'd just need to be sure that it's not causing a regression in other parts of the app.

**Comment by parasharrajat** [comment]
Thanks for the thoughts @marcochavezf. I agree after analysis I can say it is overcomplicated to work on a single context menu for all cases in this issue. For now, multiple context menus are working fine and not causing issues. So my focus will be to fix the root cause of this issue. 

There is still some part left from my side to finish the review. I will keep you posted.

**Comment by parasharrajat** [comment]
I will get back to it. Had some delays to be some setup issues on my Mac.

**Comment by arielgreen** [comment]
@parasharrajat update please

**Comment by parasharrajat** [comment]
Yeah, It seems like I won't get the time to review this today but surely tomorrow.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by parasharrajat** [comment]
Currently being reviewed.

**Comment by parasharrajat** [comment]
Ok, let's discuss your proposal @b1tjoy. There are a few regression happening here but we can ignore those for now. 

Questions: 
1. Why do we need to show popover on the `PreRenderer`? 
2. You have shown the code but didn't explain why for each piece. 

> Long press on code block comment will failed, because we use <View onStartShouldSetResponder={() => true}> in component PreRenderer

It works for me. Reason for adding this https://github.com/Expensify/App/issues/7497#issuecomment-1107902715.

3. Did you add the custom popover opening logic to imageRender as removing the `LongPressGesturehandler` will break the longpress on the images? 
4. `showPopover` is exposed via context because report data is not available at `Imagerenderer`.
5. Exposing `Showpopover` does not look robust. I have discussed this in the past to use a ContextMenu but that was rejected at that time. I will open a new discussion to see if that is something that can be done now. 

May be export `ReportActionItemData` via context and then use it for the contextMenu.

**Comment by parasharrajat** [comment]
> showPopover depends on other property member of ReportActionItem implicitly, I agree with you it's not robust. I exposed necessary data through Context API in my previous proposal, will try the ReportActionItemData way in my updated proposal.

Please do compare the rendering costs. IMO, ReportActionItemData context will trigger more renders. I believe the current will not as method definition will never change.

Let me know when you have updated your proposal.

**Comment by arielgreen** [comment]
@parasharrajat please review.

**Comment by kevinksullivan** [comment]
Just checking in. @parasharrajat when are we looking at a review for this one?

**Comment by JmillsExpensify** [comment]
@marcochavezf To help clear out the extensive backlog of /App bugs, we're putting the spotlight on all bugs older than 4 weeks old. To help unblock the roadmap and get our bug pipeline back in equilibrium, can you:

 - Decide whether any proposals currently meet our guidelines and can be approved as-is
 - For any that can't, please take this issue internal and treat it as one of your highest priorities
 - If you have any questions, don't hesitate to start a discussion in #bug-zero

Thanks everyone!

**Comment by parasharrajat** [comment]
Next update in 10 hours.

**Comment by parasharrajat** [comment]
I am not [feeling well](https://expensify.slack.com/archives/C02NK2DQWUX/p1668778734756069) so please hold for a bit more. Thanks.

**Comment by parasharrajat** [comment]
Ok, let's go with @b1tjoy [proposal](https://github.com/Expensify/App/issues/8311#issuecomment-1312527754). I am fine with using Context API. 

cc: @marcochavezf 

:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by marcochavezf** [comment]
Perfect, if there's another edge case or regression let's fix them in the PR 👍🏽

**Comment by arielgreen** [comment]
@b1tjoy @parasharrajat @mateusbra Upwork posting, please apply https://www.upwork.com/jobs/~01a8294a9c054ef83d

**Comment by mateusbra** [comment]
applyed 👍

**Comment by girraj-ch** [comment]
Hello @MelvinBot 
My Name Girraj Choudhary from India. I have 15+ years experience in UX/UI.

My best skills are HTML5, CSS3, SCSS, Tailwinds CSS, Bootstrap, JavaScript, jQuery, React JS, Next js, Vue JS,  Angular, Node JS, NPM, Adobe XD, Adobe Photoshop Figma, Github.

I am expert in UX & UI. if you have any UX/UI requirements so Please give me one chance. I will provide good quality work.  
Upwork
https://www.upwork.com/freelancers/~01f35da3087b41933a
Freelancer
https://www.freelancer.com/u/girraj12

I will wait for your response 

Thank you
Girraj

**Comment by workandplayhard** [comment]
Hi, @MelvinBot 

My name is Gabriel Zin from the Netherlands. I have 8+ years experience of web development, mainly focused on front-end.

My powerful skills are React, Vue.js, AngularJS, NodeJS, Next.js, Nuxt.js, Quasar Framework, AWS, Docker, etc.

I'd really want to jump right into this opportunity and I'm sure that I can contribute to build and fix your project with my solid and powerful skills.

Feel free to message me to discuss more details through email or chat.

My email address is gabryalzin@gmail.com.

Looking forward to hearing from you soon.

Best regards
Gabriel

**Comment by arielgreen** [comment]
@parasharrajat is reviewing this today.

**Comment by parasharrajat** [comment]
Link to Slack thread https://expensify.slack.com/archives/C01GTK53T8Q/p1669825044889799

**Comment by workandplayhard** [comment]
Ok. I'll join.

**Comment by girraj-tm** [comment]
@parasharrajat  I can't join.

**Comment by JmillsExpensify** [comment]
Is it possible to get someone else to review this in the meantime? I want to be sensitive to wrapping up the PR review and getting this merged.

**Comment by marcochavezf** [comment]
I'm going to review it today 👍🏽

**Comment by JmillsExpensify** [comment]
Nice one! I don't think you were able to get to this today, right? If I'm not missing something, let's close the loop tomorrow and get this issue moving again.

**Comment by marcochavezf** [comment]
Yeah, I haven't finished testing it out on all platforms, but hopefully tomorrow it will be done.

**Comment by marcochavezf** [comment]
Waiting for @b1tjoy to address the changes requested, then I will continue with the testing.

**Comment by marcochavezf** [comment]
Thanks for catching that @b1tjoy, let us know when it's done 👍🏽

**Comment by marcochavezf** [comment]
Not overdue, I'm reviewing it.

**Comment by arielgreen** [comment]
Touching base with @marcochavezf [here](https://expensify.slack.com/archives/C01SKUP7QR0/p1674587738007959).

**Comment by marcochavezf** [comment]
Completed the BugZero checklist 👍🏽

**Comment by arielgreen** [comment]
still discussing

**Comment by arielgreen** [comment]
Payments issued.

No regression test needed, it's already covered [here](https://expensify.testrail.io/index.php?/cases/view/1971014&group_by=cases:section_id&group_order=asc&display_deleted_cases=0&group_id=229062).
